staging/src/k8s.io/client-go/util/consistencydetector/list_data_consistency_detector.go

```diff
@@ -0,0 +1,70 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package consistencydetector
+
+import (
+	"context"
+	"os"
+	"strconv"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+var dataConsistencyDetectionForListAgainstCacheEnabled = false
+
+func init() {
+	dataConsistencyDetectionForListAgainstCacheEnabled, _ = strconv.ParseBool(os.Getenv("KUBE_LIST_AGAINST_CACHE_INCONSISTENCY_DETECTOR"))
+}
+
+// CheckListAgainstCacheDataConsistencyIfRequested performs a data consistency check only when
+// the KUBE_LIST_AGAINST_CACHE_INCONSISTENCY_DETECTOR environment variable was set during a binary startup
+// for requests served from the watch-cache.
+//
+// The consistency check is meant to be enforced only in the CI, not in production.
+// The check ensures that data retrieved by a list api call against the watch-cache
+// is exactly the same as data received by the list api call against etcd.
+//
+// Note that this function will panic when data inconsistency is detected.
+// This is intentional because we want to catch it in the CI.
+//
+// Note that this function doesn't examine the ListOptions to determine
+// if the original request has hit the cache because it would be challenging
+// to maintain consistency with the server-side implementation.
+// For simplicity, we assume that the first request retrieved data from
+// the cache (even though this might not be true for some requests)
+// and issue the second call to get data from etcd for comparison.
+func CheckListAgainstCacheDataConsistencyIfRequested[T runtime.Object](ctx context.Context, identity string, listItemsFn ListFunc[T], optionsUsedToReceiveList metav1.ListOptions, receivedList runtime.Object) {
+	if !dataConsistencyDetectionForListAgainstCacheEnabled {
+		return
+	}
+	checkListAgainstCacheDataConsistencyIfRequestedInternal(ctx, identity, listItemsFn, optionsUsedToReceiveList, receivedList)
+}
+
+func checkListAgainstCacheDataConsistencyIfRequestedInternal[T runtime.Object](ctx context.Context, identity string, listItemsFn ListFunc[T], optionsUsedToReceiveList metav1.ListOptions, receivedList runtime.Object) {
+	receivedListMeta, err := meta.ListAccessor(receivedList)
+	if err != nil {
+		panic(err)
+	}
+	rawListItems, err := meta.ExtractListWithAlloc(receivedList)
+	if err != nil {
+		panic(err) // this should never happen
+	}
+	lastSyncedResourceVersion := receivedListMeta.GetResourceVersion()
+	CheckDataConsistency(ctx, identity, lastSyncedResourceVersion, listItemsFn, optionsUsedToReceiveList, func() []runtime.Object { return rawListItems })
```

Comment:
I think this may fail in some cases - IIRC when Continuation is set, RV can't be set.
We need to double check some things there.

Reply:
so we actually need some bits from the previous function (`wasListRequestServedFromStorage`).

Reply:
I don't think so

I think you need to modify CheckDataConsistency to set RV/RVMatch only if continuation is unset.
Need to look deeper, but sth like that should do the job.

Reply:
opened https://github.com/kubernetes/kubernetes/pull/125166